analysis/src/python/data_analysis/template_errors/search.py

```diff
@@ -0,0 +1,113 @@
+    Represent issue as a pair <origin_class, line>.
+    """
+    issues = submission[SubmissionColumns.RAW_ISSUES.value]
+    return [(issue.origin_class, submission[SubmissionColumns.CODE.value][issue.line_no - 1]) for issue in issues]
+
+
+equal_lines: Callable[[str, str], bool] = lambda x, y: x == y
+
+
+def equal_issues(issue: Tuple[str, str], other_issue: Tuple[str, str]) -> bool:
+    return issue[0] == other_issue[0] and equal_lines(issue[1], other_issue[1])
+
+
+def get_repetitive_issues(issues_by_attempt: List[List[Tuple[str, str]]]) -> List[Tuple[str, str]]:
+    """
+    Find issues that occur in every attempt of the series issues_by_attempt.
+    """
+
+    def is_in_previous_issues(issue_line: Tuple[str, str],
+                              previous_issues_lines: List[Tuple[str, str]]) -> bool:
+        for pr_issue_line in previous_issues_lines:
+            if equal_issues(issue_line, pr_issue_line):
+                return True
+        return False
+
+    prev_issues = issues_by_attempt[0]
+
+    for attempt in range(1, len(issues_by_attempt)):
+        issues = issues_by_attempt[attempt]
+        prev_issues = [issue for issue in issues if is_in_previous_issues(issue, prev_issues)]
+
+    return prev_issues
+
+
+def get_candidate_issues(df_submissions: pd.DataFrame) -> pd.Series:
+    """
+    Rank issues by frequency of occurrence in all attempts of the series.
+    """
+    candidate_issues = []
+    groups = df_submissions[SubmissionColumns.GROUP.value].unique()
+
+    for group_no in groups:
+        df_issues = df_submissions[df_submissions[SubmissionColumns.GROUP.value] == group_no] \
+            .sort_values(by=SubmissionColumns.ATTEMPT.value) \
+            .apply(lambda x: submission_to_issue_code_pairs(x), axis=1)
+        candidate_issues += get_repetitive_issues(df_issues.to_list())
+
+    ranking = pd.DataFrame(candidate_issues).value_counts()
+    ranking /= len(groups)
+    return ranking
+
+
+def search(submissions_path: str, result_path: str, n: int):
+    """
+    Get n most frequently uncorrected issues for every step_id in submissions_path and write them to result_path.
+    """
+    df_submissions = read_df(submissions_path)
+
+    df_submissions = df_submissions[[SubmissionColumns.ID.value,
+                                     SubmissionColumns.STEP_ID.value,
+                                     SubmissionColumns.CODE.value,
+                                     SubmissionColumns.GROUP.value,
+                                     SubmissionColumns.ATTEMPT.value,
+                                     SubmissionColumns.RAW_ISSUES.value]]
+
+    # Parsing raw issues
+    df_submissions[SubmissionColumns.RAW_ISSUES.value] = \
+        df_submissions[SubmissionColumns.RAW_ISSUES.value].map(lambda x: json.loads(x, cls=RawIssueDecoder))
+    # Splitting code to lines
+    df_submissions[SubmissionColumns.CODE.value] = \
+        df_submissions[SubmissionColumns.CODE.value].map(lambda x: x.split('\n'))
```

Comment:
Maybe should we use [`os.linesep`](https://docs.python.org/3/library/os.html#os.linesep)?